File: cpp/search2dMatrix2.cpp
#include <iostream>
#include <ostream>
#include <vector>
class Solution {
public:
  bool searchMatrix(std::vector<std::vector<int>> &matrix, int target) {
    int x = matrix.size();
    int y = matrix[0].size();

    int leftX = 0;
    int rightX = x - 1;

    while (leftX < rightX) {
      int middleX = leftX + (rightX - leftX) / 2;
      if (target >= matrix[middleX][0] && target <= matrix[middleX][y - 1]) {

        int leftY = 0;
        int rightY = y - 1;

        while (leftY < rightY) {
          int middleY = leftY + (rightY - leftY) / 2;
          if (target > matrix[middleX][middleY]) {
            leftY = middleY + 1;
          } else if (target < matrix[middleX][middleY]) {
            rightY = middleY - 1;
          } else if (target == matrix[middleX][middleY]) {
            return true;
          } else {
            return false;
          }
        }
      } else if (target < matrix[middleX][0]) {
        rightX = middleX - 1;
      } else if (target > matrix[middleX][y - 1]) {
        leftX = middleX + 1;
      } else {
        std::cout << "Here" << std::endl;
        return false;
      }
    }
    std::cout << "Here2" << std::endl;
    return false;
  }
};

File: python/search_2d_matrix2.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROW_LEN = len(matrix[0]) - 1

        l_row = 0
        r_row = len(matrix) - 1

        while l_row <= r_row:
            m_row = (l_row + r_row) // 2
            if target > matrix[m_row][ROW_LEN]:
                l_row = m_row + 1
            elif target < matrix[m_row][0]:
                r_row = m_row - 1
            else:
                break

        m_row = (l_row + r_row) // 2
        l, r = 0, len(matrix[0]) - 1

        while l <= r:
            m = (l + r) // 2
            if target > matrix[m_row][m]:
                l = m + 1
            elif target < matrix[m_row][m]:
                r = m - 1
            else:
                return True

        return False
